src/client/datascience/jupyter/kernels/codeExecution.ts

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { nbformat } from '@jupyterlab/coreutils';
+import { IDisposable, IDisposableRegistry } from '../../../common/types';
+import { noop } from '../../../common/utils/misc';
+import { IDataScienceErrorHandler, IJupyterSession, INotebook } from '../../types';
+import { disposeAllDisposables } from '../../../common/helpers';
+import { createDeferred } from '../../../common/utils/async';
+import * as uuid from 'uuid/v4';
+
+export class CodeExecutionFactory {
+    constructor(
+        private readonly errorHandler: IDataScienceErrorHandler,
+        private readonly disposables: IDisposableRegistry
+    ) {}
+
+    public create(code: string) {
+        // eslint-disable-next-line @typescript-eslint/no-use-before-define
+        return CodeExecution.fromCode(code, this.errorHandler, this.disposables);
+    }
+}
+
+/**
+ * Responsible for execution of an individual cell and manages the state of the cell as it progresses through the execution phases.
+ * Execution phases include - enqueue for execution (done in ctor), start execution, completed execution with/without errors, cancel execution or dequeue.
+ *
+ * WARNING: Do not dispose `request: Kernel.IShellFuture` object.
+ * Even after request.done & execute_reply is sent we could have more messages coming from iopub.
+ * Further details here https://github.com/microsoft/vscode-jupyter/issues/232 & https://github.com/jupyter/jupyter_client/issues/297
+ *
+ */
+export class CodeExecution implements IDisposable {
```

Comment:
It seems weird that we'd need an entirely new class just to execute a string instead of a notebook cell. Could CellExecution not use this instead?

Reply:
Gets way too messy (lots of if/else), i personally think this is cleaner than trying to get a single class to do both.
Also i don't think execute hidden will require execution of code that has widgets & other complexities.

Reply:
No not a single class. The other class would use this one. This one knows only about code. 

The other one uses this class internally to actually do an execute. It just listens to the promise and updates the cell outputs on the cell. 

Otherwise we end up with the same code doing iopub listening and the potential for having to fix things in two spots if something is broken.

Reply:
>  The other class would use this one. This one knows only about code.

I'm still not convinced, this is simple and small, it just gets messy. We don't need widgets, etc, hence this code is sufficient as is.

> code doing iopub listening and the potential for having to fix things in two

I'd perfer to refactor when that happens, right now this feels much simpler to me, but thats just me.